### PR TITLE
introduce a proper trace context

### LINKED ISSUE #124073: K8s trace context for APIServer is incorrect
### What happened?

In Kubernetes release 1.28, the tracing structure for the API server is depicted as follows:

image

Within the API server's code structure, the spans "List(recursive=true) etcd3" and "SerializeObject" are expected to be nested under the "List" span. However, "SerializeObject" appears parallel to "List", which is not the intended behavior.

staging/src/k8s.io/apiserver/pkg/endpoints/handlers/get.go
```golang
func ListResource(r rest.Lister, rw rest.Watcher, scope *RequestScope, forceWatch bool, minRequestTimeout time.Duration) http.HandlerFunc {
	return func(w http.ResponseWriter, req *http.Request) {
		ctx := req.Context()
		// For performance tracking purposes.

                 // CasperLiu: this span is for high level - "List"
		ctx, span := tracing.Start(ctx, "List", traceFields(req)...)

		namespace, err := scope.Namer.Namespace(req)
		if err != nil {
			scope.err(err, w, req)
			return
		}

                 // omit...
		span.AddEvent("About to List from storage")
                 // CasperLiu: span inside below function is for low level - "List(recursive=true) etcd3"
		result, err := r.List(ctx, &opts)
		if err != nil {
			scope.err(err, w, req)
			return
		}
		span.AddEvent("Listing from storage done")
		defer span.AddEvent("Writing http response done", attribute.Int("count", meta.LenList(result)))
                 // CasperLiu: span inside below function is for low level - "SerializeObject"
		transformResponseObject(ctx, scope, req, w, http.StatusOK, outputMediaType, result)
	}
}
```

### What did you expect to happen?

Within the API server's code structure, the spans "List(recursive=true) etcd3" and "SerializeObject" are expected to be nested under the "List" span. 

### How can we reproduce it (as minimally and precisely as possible)?

1. Configure the Kubernetes API server to enable tracing and direct the emitted spans to a backend. 
2. Illustrate the resulting trace flow. 
3. Invoke a quorum read LIST request for Kubernetes resources; within the trace flow, you may identify instances of incorrect trace context.

### Anything else we need to know?

_No response_

### Kubernetes version

<details>

```
kuberenetes version:
1.28.3
```

</details>

### Cloud provider

<details>
Alibaba Cloud Kubernetes (ACK) 
</details>

### OS version

<details>

Unrelated to the OS version

</details>

### Install tools

<details>
No need 
</details>

### Container runtime (CRI) and version (if applicable)

<details>
Unrelated to the CRI version

</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>

Unrelated to the plugins version

</details>

### PR DESCRIPTION
**Pull request by carlory**
#### What type of PR is this?

/kind bug

#### What this PR does / why we need it:

In Kubernetes release 1.28, the tracing structure for the API server is depicted as follows:

image

Within the API server's code structure, the spans "List(recursive=true) etcd3" and "SerializeObject" are expected to be nested under the "List" span. However, "SerializeObject" appears parallel to "List", which is not the intended behavior.

staging/src/k8s.io/apiserver/pkg/endpoints/handlers/get.go

```go
func ListResource(r rest.Lister, rw rest.Watcher, scope *RequestScope, forceWatch bool, minRequestTimeout time.Duration) http.HandlerFunc {
	return func(w http.ResponseWriter, req *http.Request) {
		ctx := req.Context()
		// For performance tracking purposes.

                 // CasperLiu: this span is for high level - "List"
		ctx, span := tracing.Start(ctx, "List", traceFields(req)...)

		namespace, err := scope.Namer.Namespace(req)
		if err != nil {
			scope.err(err, w, req)
			return
		}

                 // omit...
		span.AddEvent("About to List from storage")
                 // CasperLiu: span inside below function is for low level - "List(recursive=true) etcd3"
		result, err := r.List(ctx, &opts)
		if err != nil {
			scope.err(err, w, req)
			return
		}
		span.AddEvent("Listing from storage done")
		defer span.AddEvent("Writing http response done", attribute.Int("count", meta.LenList(result)))
                 // CasperLiu: span inside below function is for low level - "SerializeObject"
		transformResponseObject(ctx, scope, req, w, http.StatusOK, outputMediaType, result)
	}
}
```

#### Which issue(s) this PR fixes:

Fixes https://github.com/kubernetes/kubernetes/issues/124073

#### Special notes for your reviewer:

Thank @CasperLiu but your PR https://github.com/kubernetes/kubernetes/pull/124079 is inactive and will be closed by rebot. so I am taking over it and adding you as a co-author. now, it has been changed as @jpbetz suggested.

If you want to continue your PR, please let me know, and feel free to close my PR. Thanks again.

#### Does this PR introduce a user-facing change?

```release-note
Fix the wrong hierarchical structure for the child span and the parent span (i.e. `SerializeObject` and `List`). In the past, some children's spans appeared parallel to their parents.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Review by dashpole — APPROVED** [review]
/lgtm

**Comment by cici37** [comment]
/assign @sttts 
for approval. :)
/triage accepted

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/get.go [hunk 1]
```diff
@@ -171,6 +171,7 @@ func ListResource(r rest.Lister, rw rest.Watcher, scope *RequestScope, forceWatc
 		ctx := req.Context()
 		// For performance tracking purposes.
 		ctx, span := tracing.Start(ctx, "List", traceFields(req)...)
+		req = req.WithContext(ctx)
```
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/endpoints/handlers/get.go:174]: we don't have this for any other verb. Is it only needed for watch? From the PR description it does not look like. Why don't we add that everywhere to get proper serialization spans?

Comment (carlory) [reply staging/src/k8s.io/apiserver/pkg/endpoints/handlers/get.go:175]: @sttts sorry, I forgot to add it for other verbs. updated.

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/responsewriters/writers.go [hunk 2]
```diff
@@ -98,6 +98,7 @@ func SerializeObject(mediaType string, encoder runtime.Encoder, hw http.Response
 		attribute.String("protocol", req.Proto),
 		attribute.String("mediaType", mediaType),
 		attribute.String("encoder", string(encoder.Identifier())))
+	req = req.WithContext(ctx)
```
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/endpoints/handlers/responsewriters/writers.go:101]: this one is not needed, is it? `req` is not used anywhere below.

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/create.go [hunk 3]
```diff
@@ -55,6 +55,7 @@ func createHandler(r rest.NamedCreater, scope *RequestScope, admit admission.Int
 		ctx := req.Context()
 		// For performance tracking purposes.
 		ctx, span := tracing.Start(ctx, "Create", traceFields(req)...)
+		req = req.WithContext(ctx)
```
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/endpoints/handlers/create.go:58]: @dashpole these changes look good?

Comment (sttts) [reply staging/src/k8s.io/apiserver/pkg/endpoints/handlers/create.go:58]: nvmd. @dashpole lgmt'ed the change to the list handler before.

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/responsewriters/writers.go (hunk 2 above)
Comment (carlory) [reply staging/src/k8s.io/apiserver/pkg/endpoints/handlers/responsewriters/writers.go:101]: it is used in https://github.com/kubernetes/kubernetes/blob/99ff62e87a2d881f4839159d33519a49d6eee5d9/staging/src/k8s.io/apiserver/pkg/endpoints/handlers/responsewriters/writers.go#L106

**Pull request merged**